[WIP] Introduce new ChannelId struct

optout21 opened the pull request:
Fixes #2408 . Introduces a new `ChannelId` struct, enclosing the 32-byte data, and specific constructors, for normal funding-tx-based and for temporary channel IDs.

TODO:
- [ ] Determine how it affects APIs, uses, backward-compatibility
- [ ] Tests for ChannelId serialization/deserialization
- [ ] New macro log_channel_id!()
- [ ] Review hints

dunxen commented:
This overall looks like an improvement, but I'd like to maybe see what others had to say about types for different IDs like

* temporary channel IDs for V1 unfunded channels
* channel IDs for V1 funded channels with known funding txids
* temporary channel IDs (revocation basepoint XORed with zero) for V2 channels for `open_channel2` and `accept_channel2`
* channel IDs (both party revocation basepoints XORed) for V2 channels after `accept_channel2`.

For me currently I don't see a huge benefit in splitting all these cases into distinct types or even enum variants. So I think this PR is all we need so that we at least have type safety for channel ID bytes vs hash bytes for example.

optout21 commented on lightning/src/chain/transaction.rs:
```diff
@@ -56,13 +57,9 @@ pub struct OutPoint {
-		res[31] ^= ((self.index >> 0) & 0xff) as u8;
-		res
+	/// Convert an `OutPoint` to a lightning channel id. Alternatively use ChannelId::from_funding_outpoint().
+	pub fn to_channel_id(&self) -> ChannelId {
```
This method existed in OutPoint, and AFAIK OutPoint cannot be used to generate other type of ChannelId, so I did not change it now, thought it would make sense.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -603,6 +603,83 @@ impl_writeable_tlv_based!(PendingChannelMonitorUpdate, {
+
+impl ChannelId {
+	/// Create channel ID based on a funding TX outpoint
+	pub fn from_funding_outpoint(funding_tx_outpoint: &OutPoint) -> Self {
```
Renamed.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -603,6 +603,83 @@ impl_writeable_tlv_based!(PendingChannelMonitorUpdate, {
+	}
+
+	/// Serialization helper
+	pub fn serialized_length(&self) -> usize {
```
Indeed, thx

optout21 commented on lightning-invoice/src/utils.rs:
```diff
@@ -627,7 +627,7 @@ where
 	for channel in channels.into_iter().filter(|chan| chan.is_channel_ready) {
 		if channel.get_inbound_payment_scid().is_none() || channel.counterparty.forwarding_info.is_none() {
-			log_trace!(logger, "Ignoring channel {} for invoice route hints", log_bytes!(channel.channel_id));
+			log_trace!(logger, "Ignoring channel {} for invoice route hints", log_bytes!(channel.channel_id.bytes()[..]));
```
Commits squashed

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -603,6 +603,73 @@ impl_writeable_tlv_based!(PendingChannelMonitorUpdate, {
 	(0, update, required),
 });
 
+/// A unique 32-byte identifier for a channel.
```
Commits reorganized; first commit only adds the new struct, the second adapts to it.

optout21 commented:
Review comments addressed. commits regrouped (into 2 commits) as suggested.
Only outstanding issue is where to have the `ChannelId` declaration -- leave in `channel.rs`, in `mod.rs`, or in new, own mod.

jbesraa commented on lightning/src/ln/channel.rs:
```diff
@@ -603,6 +603,83 @@ impl_writeable_tlv_based!(PendingChannelMonitorUpdate, {
+/// A unique 32-byte identifier for a channel.
+/// This is not exported to bindings users as we just use [u8; 32] directly
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct ChannelId {
```
yea I think adding this to new file makes more sense

dunxen reviewed (commented):
LGTM with moving the `ChannelId` stuff to its own mod.

optout21 commented:
Moved ChannelId to its own new module (`channel_id.rs`); but need to clean up the commits

dunxen commented on lightning/src/ln/channel_id.rs:
```diff
@@ -0,0 +1,143 @@
+use core::fmt;
+use core::ops::Deref;
+
+
```
nit: there's an extra line here

dunxen commented:
Oops sorry, yeah looks good for squash

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -7554,6 +7561,47 @@ mod tests {
+	fn test_channel_id_new_from_data() {
+		let data: [u8; 32] = [2; 32];
+		let channel_id = ChannelId::from_bytes(data.clone());
+		assert_eq!(*channel_id.bytes(), data);
+	}
+
+	#[test]
+	fn test_channel_id_from_funding_tx2() {
+		let channel_id = ChannelId::v1_from_funding_txid(&[2; 32], 1);
+		assert_eq!(channel_id.to_hex(), "0202020202020202020202020202020202020202020202020202020202020203");
+	}
+
+	#[test]
+	fn test_channel_id_equals() {
+		let channel_id11 = ChannelId::v1_from_funding_txid(&[2; 32], 2);
+		let channel_id12 = ChannelId::v1_from_funding_txid(&[2; 32], 2);
+		let channel_id21 = ChannelId::v1_from_funding_txid(&[2; 32], 42);
+		assert_eq!(channel_id11, channel_id12);
+		assert_ne!(channel_id11, channel_id21);
+	}
+
+	#[test]
+	fn test_channel_id_write_read() {
+		let data: [u8; 32] = [2; 32];
+		let channel_id = ChannelId::from_bytes(data.clone());
+
+		let mut w = test_utils::TestVecWriter(Vec::new());
+		channel_id.write(&mut w).unwrap();
+
+		let channel_id_2 = ChannelId::read(&mut io::Cursor::new(&w.0)).unwrap();
+		assert_eq!(channel_id_2, channel_id);
+		assert_eq!(channel_id_2.bytes(), &data);
+	}
+
+	#[test]
+	fn test_channel_id_display() {
+		let channel_id = ChannelId::from_bytes([42; 32]);
+		assert_eq!(format!("{}", &channel_id), "2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a");
+	}
+
```
ChannelId moved to new module; together with its unit tests.

optout21 commented:
Per-commit check failed, fixed (moved last fix in first commit)

optout21 commented:
This PR touches many logging lines, so it frequently gets in conflict with current main...  I keep rebasing+resolving conflicts, but it would be nice to get more approvals ;)

optout21 commented on lightning/src/ln/channel_id.rs:
```diff
@@ -0,0 +1,144 @@
+/// The variety (context) is not stored, it is relevant only at creation.
+/// This is not exported to bindings users as we just use [u8; 32] directly
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct ChannelId {
```
Valid point! It would simplify the change.
I follow the opinion of the maintainers, as you have a much better feel of the overall style/approach of the whole project than me (and also of Rust probably).
My driver was that I find `id.0` less readable/typesafe than `id.bytes()` -- but this is subjective, and that it is easier to add fields once creation is behind constructors -- but there's little chance foreseen and it can be changed later if needed.
I think I'll rework as suggested.

optout21 replied:
`.bytes()` has not much value in that case; I would keep the `from_bytes()` generic ctor, to suggest that constructor should be used, and have a place to insert the comment to use the other specific ctors whenever possible :)

optout21 commented:
Simplified as suggested (@TheBlueMatt): changed struct into a pub tuple; removed `.bytes()` (use `.0` instead).
Also found some instances in tests where temp ID is generated, but `from_bytes()` was used.
Also some rustdoc formatting.

optout21 commented:
2 CI builds fail, likely due to #2527 . 2 approvals.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4752,14 +4755,14 @@ impl<SP: Deref> Channel<SP> where
 				for inp in tx.input.iter() {
 					if inp.previous_output == funding_txo.into_bitcoin_outpoint() {
-						log_info!(logger, "Detected channel-closing tx {} spending {}:{}, closing channel {}", tx.txid(), inp.previous_output.txid, inp.previous_output.vout, log_bytes!(self.context.channel_id()));
+						log_info!(logger, "Detected channel-closing tx {} spending {}:{}, closing channel {}", tx.txid(), inp.previous_output.txid, inp.previous_output.vout, &self.context.channel_id());
```
Thx for the explainer, I was not aware of it. If that was not the case, the worse that could happen is a 32-byte value copy instead of a pointer copy, which is not nice but not catastrophic.

optout21 commented on lightning/src/ln/payment_tests.rs:
```diff
@@ -16,10 +16,10 @@ use crate::chain::channelmonitor::{ANTI_REORG_DELAY, HTLC_FAIL_BACK_BUFFER, LATE
 use crate::chain::transaction::OutPoint;
 use crate::events::{ClosureReason, Event, HTLCDestination, MessageSendEvent, MessageSendEventsProvider, PathFailure, PaymentFailureReason, PaymentPurpose};
-use crate::ln::channel::EXPIRE_PREV_CONFIG_TICKS;
+use crate::ln::channel::{EXPIRE_PREV_CONFIG_TICKS};
```
Hmm, typo, after subsequent add and delete.
